File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        charSet = {')' : '(', ']' : '[', '}' : '{'}
        st = []
        for i in range(len(s)):
            if s[i] in charSet:
                if not st or st[-1] != charSet[s[i]]:
                    return False
                st.pop()
            else:
                st.append(s[i])

        return not st 

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> pair = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        stack<char> st;

        for (int i = 0; i < s.size(); ++i){
            if (!st.empty() && pair[s[i]] == st.top()){
                st.pop();
            }
            else{
                st.push(s[i]);
            }
        } 

        return st.empty();
    }
};